### DISCUSSION TITLE
Upsert performance

### DISCUSSION THREAD

**Discussion by pkese** [duckdb/duckdb#5987] [General]
I'm quite amazed that UPSERT got implemented https://github.com/duckdb/duckdb/pull/5866  
Big kudos to the authors.

I'd like to ask what are the performance implications or other costs (e.g. write amplification) when using upserts?

Columnar databases are known not to be very efficient when overwriting data or doing small inserts.

How are upserts implemented on the database (storage layer) level?

**Comment by Tishj** [comment]
Hey there, thanks so much for the interest in UPSERT!

I unfortunately don't have any in-depth performance insights to share, the main focus of that PR has been to add the functionality, and later worry about optimizing this.

Having said that, these are some internal details:

Currently, we have to disable parallel insertion when doing a UPSERT with the DO UPDATE action.
This is done to make sure the behavior is consistent in the event of a double update.
This was one of the more poorly documented behaviors of UPSERT in my opinion, Postgres (and we do as well) throws an error when you attempt to do an update of the same row, in the same statement.
SQLite only performs the first of the updates, and does not throw an error.

To support this we need to check if a given row is already updated, which would be a big bottleneck in the parallel case with a naive implementation.

Other than that, this is the general flow of the UPSERT for a chunk:

Check constraints, record the conflicts
Fetch the rows of the existing table that were hit by a conflict (if they are required by any of the conditions/set expressions)

Check the initial conflict condition (ON CONFLICT (...) WHERE <condition>)
If the action is DO NOTHING, we're done here

Add which rows we are going to update (this is where the check for double updates happens)
Check the optional DO UPDATE condition, which just filters out which rows are going to be hit by the update
Then perform the update



Some relatively obvious candidates for optimization here:
Delay the fetch of the existing tuples, since these are pretty much random access point queries, it would be faster to group them and fetch them per segment/rowgroup

And then to both fix the bottleneck and improve performance, we probably want to delay the updates themselves too, and only collect which rows are set for update, so we can keep this all parallel, then during combine check for the double-updates
Only once that is done, perform the update
